### PR TITLE
8173585: Intrinsify StringLatin1.indexOf(char)

### PR DESCRIPTION
**Pull request by jasontatton-aws**
This is an implementation of the indexOf(char) intrinsic for StringLatin1 (1 byte encoded Strings). It is provided for x86 and ARM64. The implementation is greatly inspired by the indexOf(char) intrinsic for StringUTF16. To incorporate it I had to make a small change to StringLatin1.java (refactor of functionality to intrisified private method) as well as code for C2.

Submitted to: hotspot-compiler-dev and core-libs-dev as this patch contains a change to hotspot and java/lang/StringLatin1.java

https://bugs.openjdk.java.net/browse/JDK-8173585

Details of testing: 
============
I have created a jtreg test “compiler/intrinsics/string/TestStringLatin1IndexOfChar” to cover this new intrinsic. Note that, particularly for the x86 implementation of the intrinsic, the code path taken is dependent upon the length of the input String. Hence the test has been designed to cover all these cases. In summary they are:

- A “short” string of < 16 characters.
- A SIMD String of 16 – 31 characters.
- A AVX2 SIMD String of 32 characters+.

Hardware used for testing: 
-----------------------------

- Intel Xeon CPU E5-2680 (JVM did not recognize this as having AVX2 support) • Intel i7 processor (with AVX2 support).
- AWS Graviton 2 (ARM 64 processor).

I also ran; ‘run-test-tier1’ and ‘run-test-tier2’ for: x86_64 and aarch64.

Possible future enhancements: 
====================
For the x86 implementation there may be two further improvements we can make in order to improve performance of both the StringUTF16 and StringLatin1 indexOf(char) intrinsics:

1. Make use of AVX-512 instructions.
2. For “short” Strings (see below), I think it may be possible to modify the existing algorithm to still use SSE SIMD instructions instead of a loop.

Benchmark results:
============
**Without** the new StringLatin1 indexOf(char) intrinsic:

| Benchmark  | Mode | Cnt | Score | Error | Units |
| ------------- | ------------- |------------- |------------- |------------- |------------- |
| IndexOfBenchmark.latin1_mixed_char | avgt | 5 | **26,389.129** | ± 182.581 | ns/op |
| IndexOfBenchmark.utf16_mixed_char | avgt | 5 | 17,885.383 | ± 435.933  | ns/op |

**With** the new StringLatin1 indexOf(char) intrinsic:

| Benchmark  | Mode | Cnt | Score | Error | Units |
| ------------- | ------------- |------------- |------------- |------------- |------------- |
| IndexOfBenchmark.latin1_mixed_char | avgt | 5 | **17,875.185** | ± 407.716  | ns/op |
| IndexOfBenchmark.utf16_mixed_char | avgt | 5 | 18,292.802 | ± 167.306  | ns/op |

The objective of the patch is to bring the performance of StringLatin1 indexOf(char) in line with StringUTF16 indexOf(char) for x86 and ARM64. We can see above that this has been achieved. Similar results were obtained when running on ARM.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8173585](https://bugs.openjdk.java.net/browse/JDK-8173585): Intrinsify StringLatin1.indexOf(char)

### Reviewers
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/71/head:pull/71`
`$ git checkout pull/71`

### CODE REVIEW

**Comment by jasontatton-aws** [comment]
Hi Andrew,

The current indexOf char intrinsics for StringUTF16 and the new one here for StringLatin1 both use the AVX2 – i.e. 256 bit instructions, these are also affected by the frequency scaling which affects the AVX-512 instructions you pointed out. Of course in a world where all the work taking place is AVX instructions this wouldn’t be an issue but in mixed mode execution this is a problem.

However, the compiler does have knowledge of the capability of the CPU upon which it’s optimizing code for and is able to decide whether to use AVX instructions if they are supported by the CPU AND if it wouldn’t be detrimental for performance. In fact, there is a flag which one can use to interact with this: -XX:UseAVX=version.

This of course made testing this patch an interesting experience as the AVX2 instructions were not enabled on the Xeon processors which I had access to at AWS, but in the end I was able to use an i7 on my corporate macbook to validate the code.

**Comment by jasontatton-aws** [comment]
Hi everyone,

This patch is just missing a couple of reviewers... Please can someone step forward?

I think it's a fairly straightforward change.

-Jason

**Comment by jasontatton-aws** [comment]
Hi Andrew,

Thanks for coming back to me. Looking on the github [PR](https://github.com/openjdk/jdk/pull/71) nobody is tagged as a reviewer for this (perhaps this is a feature which is not being used).

> That's
> because what is actually missing is the justification he asked for. As
> Andrew pointed out the change is simple but the reason for implementing
> it is not.

There are two separate things here:
1). Justification for the change itself:
-The objective and justification for this patch is to bring the performance of StringLatin1 indexOf(char) in line with StringUTF16 indexOf(char) for x86 and ARM64. This solves the problem as raised in [JDK-8173585](https://bugs.openjdk.java.net/browse/JDK-8173585), and also on the [mailing list](http://mail.openjdk.java.net/pipermail/jdk9-dev/2017-January/005539.html).

2). Discussion around future enhancements - concerning potential use of AVX-512 instructions and a more optimal implementation for short strings.
-These would be separate JBS's I'm not advocating for/against this, they are just ideas separate from this JBS.

**Comment by jasontatton-aws** [comment]
The AVX code path represents approximately 1/6th of the patch (1/7th including the infrastructure  code around this). I don’t think we should discard the entire patch because 1/6th of the code may have unintended consequences. This is especially the case when the rest of the code has been benchmarked, with certainty, to show the desired performance improvement has been achieved.
 
Additionally, I do not see how those unintended consequences will ever be realised because the JVM has knowledge of the AVX capability of the chip it’s running on and disables the AVX2 code path for chips which suffer from the performance degradation which has been outlined in this discussion. Thus protecting us from unintended consequences. Unless we are asserting that this mechanism to globally control the use of AVX2 instructions is broken or otherwise non functional I see no reason to remove the AVX2 code. And to be consistent we would really need to look at removing all instances of AVX2 code in the JVM (of which there is quite a lot).
 
As I see it there are three ways forward:

1.  Accept the patch as is. 
2. Modify the patch to remove the AVX code path for x86, and/or any other modifications needed.
3. Discard the patch entirely.

At this point I am in favour of approach 1 but happy to accept 2 if advised that this is the right thing to do.

**Comment by adinn** [comment]
"the JVM has knowledge of the AVX capability of the chip it’s running on and disables the AVX2 code path for chips which suffer from the performance degradation which has been outlined in this discussion"

Does it? The white paper Andrew cited doesn't mention this as being specific to only some chips that implement AVX2. Can you explain where this restricted effect is documented?

Also, I assume you are referring to the code in vm_version_x86.cpp with this comment

    // Don't use AVX-512 on older Skylakes unless explicitly requested

is that correct?

**Comment by jasontatton-aws** [comment]
> Can you explain where this restricted effect is documented?

Certainly! I’ve found that determining the capability of the CPU and whether to enable AVX2 support if the chip supports it is mostly controlled in: [vm_version_x86.cpp]( https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/vm_version_x86.cpp) specifically: [get_processor_features](https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/vm_version_x86.cpp#L684-L755) and in [generate_get_cpu_info]( https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/vm_version_x86.cpp#L69-L611).

In order to test the patch comprehensively I had to track down an Intel Core i7 (I7-9750H) processor which the aforementioned code permitted AVX2 instructions for (maybe this is an error and it should not be enabled for this processor though) as most of the infrastructure I personally use here at AWS runs on Intel Xeon processors - I also tested on a E5-2680 which the JVM does not enable AVX2 for.

However, this is just the Intel side of things. When it comes to AMD I read that the AMD Zen 2 architecture, of which the current flagship: Threadripper 3990X, is based, is able to support AVX2 [without the frequency scaling]( https://www.anandtech.com/Show/Index/14525?cPage=7&all=False&sort=0&page=9&slug=amd-zen-2-microarchitecture-analysis-ryzen-3000-and-epyc-rome) which some/all(?) of the Intel chips incur. I personally don’t have access to one of these chips so I cannot confirm how it is classified in the JVM.

Also, I found when investigating this that there is actually a JVM flag which can be used to control what level of AVX is enabled: `-XX:UseAVX=version.`

**Comment by mknjc** [comment]
I really don't see the problem with using AVX for this?
As long as the used instructions all only activate license l0 the cpu don't scale the frequency at all. For AVX2 these are most of all instructions which don't use the FMA or floating point ports.
Additionally the cpu doesn't instant scale down the frequency but runs the 256 bit instructions with reduced throughput but full cpu clock until enough instructions use the avx command set.
For more information see https://stackoverflow.com/a/56861355/130541 

So as long the 512bit width instructions aren't used there should no frequency scaling happening.

**Comment by adinn** [comment]
>  >    Can you explain where this restricted effect is documented?

> Certainly! I’ve found that determining the capability of the CPU and whether to enable AVX2 support if the chip supports it is mostly controlled in: vm_version_x86.cpp specifically: get_processor_features and in generate_get_cpu_info.

Yes, I can see what the code does. I was asking where the cpu behaviour is documented independent of the code.

> In order to test the patch comprehensively I had to track down an Intel Core i7 (I7-9750H) processor which the aforementioned code permitted AVX2 instructions for (maybe this is an error and it should not be enabled for this processor though) as most of the infrastructure I personally use here at AWS runs on Intel Xeon processors - I also tested on a E5-2680 which the JVM does not enable AVX2 for.

'maybe'? The documentation Andrew provided mentioned Xeon E5 v3 which I believe is a Skylake design. However, the code I pointed you at in vm_version_x86 which claims to detect 'early Skylake' designs is only disabling AVX512 support. It still enables AVX2. Similarly, the code generates machine code to check the processor capabilities has a special check if use_evex is set (i.e. AVX3 is requested) for Skylake which disables AVX512 but does nto disable AVX2 support.

However, this is just the Intel side of things. When it comes to AMD I read that the AMD Zen 2 architecture, of which the current flagship: Threadripper 3990X, is based, is able to support AVX2 without the frequency scaling which some/all(?) of the Intel chips incur. I personally don’t have access to one of these chips so I cannot confirm how it is classified in the JVM.

Also, I found when investigating this that there is actually a JVM flag which can be used to control what level of AVX is enabled: -XX:UseAVX=version.

File: src/hotspot/share/classfile/vmSymbols.cpp [hunk 1]
```diff
@@ -292,4 +292,4 @@ vmSymbols::SID vmSymbols::find_sid(const char* symbol_name) {
   Symbol* symbol = SymbolTable::probe(symbol_name, (int) strlen(symbol_name));
   if (symbol == NULL)  return NO_SID;
   return find_sid(symbol);
-}
+}
```
Comment (simonis) [inline src/hotspot/share/classfile/vmSymbols.cpp:295]: I think it is common sense to have a newline at the end of a file.

**Comment by vnkozlov** [comment]
Based on comment by @jatin-bhateja (Intel) frequency level switchover pointed by @theRealAph  is sensitive to vector size
https://github.com/openjdk/jdk/pull/144#issuecomment-692044896

By keeping vector size less or equal to 32 bytes we should avoid it. And as I can see this intrinsic code is using 32 bytes (chars) and 16 bytes vectors: `pbroadcastb(vec1, vec1, Assembler::AVX_256bit);`

Also we never had issues with AVX2. only with AVX512 regarding performance hit:
https://bugs.openjdk.java.net/browse/JDK-8221092

I would like to see performance numbers for for all values of UseAVX flag : 0, 1, 2, 3

The usage is guarded UseSSE42Intrinsics in UseSSE42Intrinsics predicate in .ad file. Make sure to test with UseAVX=0 to make sure that some avx instructions are not mixed into non avx code. And also with UseSSE=2 (for example) to make sure shared code correctly recognize that intrinsics is not supported.

**Comment by adinn** [comment]
@vnkozlov @mknjc  @jatin-bhateja Thanks for providing the relevant details. I'm now quite content that this patch avoids any potential frequency scaling problem. I'm also glad that an explanation of why this is so is now available --  although it's not perfect that we are relying on a stackoverflow post for the full details.

**Review by simonis — CHANGES_REQUESTED** [review]
Hi Jason,

thanks for bringing String.indexOf() for latin strings up to date with the Unicode version.

Your changes look good except a few minor issues I've commented on right in the code.

I'd only like to ask you if you could possibly improve your test a little bit. As far as I understand, your search text is a consecutive sequence of "abc" characters, so you'll always find the character your searching for within the next three characters of the source text. This won't exercise the loops of your intrinsic. Maybe you can also add some test versions where the search character will be found beyond the first 32/64 characters after "fromIndex"?

File: test/hotspot/jtreg/compiler/intrinsics/string/TestStringLatin1IndexOfChar.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+import jdk.test.lib.Asserts;
+import java.util.HashMap;
+import java.util.Map;
+
+public class TestStringLatin1IndexOfChar{
+    private final static int MAX_LENGTH = 513;//future proof for AVX-512 instructions
+
+    public static void main(String[] args) throws Exception {
+        for (int i = 0; i < 100_0; ++i) {//repeat such that we enter into C2 code...
```
[the remark this answers is not in the document]
Comment (jasontatton-aws) [reply test/hotspot/jtreg/compiler/intrinsics/string/TestStringLatin1IndexOfChar.java]: I have added Tier4InvocationThreshold=200 to the unit test config in order to trigger generation earlier

**Comment by jasontatton-aws** [comment]
@simonis Thank you for the corrections, I have ammended them in the latest comit as follows:

Changes to unit test:
- main test adjusted such that Strings gennerated are much longer (up to 2048 characters) and of the form: `azaza`, `aazaazaa`, `aaazaaazaaa`, etc with `'z'` being the search character searched for. Multiple instances of the search character are included in the String in order to validate that the starting offset is correctly handleded. Results are compared to non intrinsified version of the code. Longer strings means that the looping functionality of the various paths is entered into.
- Run configurations introduced such that it checks behaviour where use of SSE and AVX instructions are restricted.
- Tier4InvocationThreshold adjusted so as to ensure C2 code iis invoked.

Other changes:
- newlines added at end of files

@vnkozlov here are the performance numbers as requested. I have included performance of the UTF16 version of the intrinsic for reference:

| UseAVX= | UseSSE= | Benchmark                         | Mode | Cnt | Score       | Error       | Units |
|---------|---------|-----------------------------------|------|-----|-------------|-------------|-------|
|         | 0       | IndexOfBenchmark.latin1_long_char | avgt | 5   | **447,493.398** | ± 4,666.386 | ns/op |
| 0       |         | IndexOfBenchmark.latin1_long_char | avgt | 5   | **104,735.941** | ± 2,484.403 | ns/op |
| 1       |         | IndexOfBenchmark.latin1_long_char | avgt | 5   | **104,342.844** | ± 2,656.343 | ns/op |
| 2       |         | IndexOfBenchmark.latin1_long_char | avgt | 5   | **61,000.418**  | ± 1,543.951 | ns/op |
| 3       |         | IndexOfBenchmark.latin1_long_char | avgt | 5   | **60,607.988**  | ± 1,466.354 | ns/op |
|         | 0       | IndexOfBenchmark.utf16_long_char  | avgt | 5   | 672,475.302 | ± 4,998.596 | ns/op |
| 0       |         | IndexOfBenchmark.utf16_long_char  | avgt | 5   | 175,521.654 | ± 7,549.094 | ns/op |
| 1       |         | IndexOfBenchmark.utf16_long_char  | avgt | 5   | 172,514.981 | ± 3,561.040 | ns/op |
| 2       |         | IndexOfBenchmark.utf16_long_char  | avgt | 5   | 120,725.748 | ± 2,004.400 | ns/op |
| 3       |         | IndexOfBenchmark.utf16_long_char  | avgt | 5   | 120,664.623 | ± 1,988.419 | ns/op |

I think the results are as expected, we see improvements in performance as the range of SSE and AVX instructions which can be used is expanded upon. Note that no improvement is observed with UseAVX=3 because there is no AVX-512 code in these intrinsics.

File: test/hotspot/jtreg/compiler/intrinsics/string/TestStringLatin1IndexOfChar.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+ * @run main/othervm -Xbatch -XX:Tier4InvocationThreshold=200 -XX:CompileThreshold=100 -XX:UseAVX=1 compiler.intrinsics.string.TestStringLatin1IndexOfChar
+ * @run main/othervm -Xbatch -XX:Tier4InvocationThreshold=200 -XX:CompileThreshold=100 -XX:UseAVX=2 compiler.intrinsics.string.TestStringLatin1IndexOfChar
+ * @run main/othervm -Xbatch -XX:Tier4InvocationThreshold=200 -XX:CompileThreshold=100 -XX:UseAVX=3 compiler.intrinsics.string.TestStringLatin1IndexOfChar
+ */
+
+package compiler.intrinsics.string;
+
+import jdk.test.lib.Asserts;
+
+public class TestStringLatin1IndexOfChar{
```
Comment (neliasso) [inline test/hotspot/jtreg/compiler/intrinsics/string/TestStringLatin1IndexOfChar.java:25]: Can you please add testing for these edge cases:
- when the search char is the first char 
- when the search char is the last char 
- when the string has length 1

Comment (jasontatton-aws) [reply test/hotspot/jtreg/compiler/intrinsics/string/TestStringLatin1IndexOfChar.java:25]: Thanks for reviewing this. I have added a new test: `findOneItem` which covers these edge cases

File: test/hotspot/jtreg/compiler/intrinsics/string/TestStringLatin1IndexOfChar.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
+/*
```
Comment (vnkozlov) [inline test/hotspot/jtreg/compiler/intrinsics/string/TestStringLatin1IndexOfChar.java:1]: Missing copyright+GPL header in new test. See other tests fro example.

File: test/micro/org/openjdk/bench/java/lang/StringIndexOfChar.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
+package org.openjdk.bench.java.lang;
```
Comment (vnkozlov) [inline test/micro/org/openjdk/bench/java/lang/StringIndexOfChar.java:1]: Again missing Copyright.

**Review by vnkozlov — CHANGES_REQUESTED** [review]
Changes seems fine but you missing Copyright + GPL header in new files.

File: test/hotspot/jtreg/compiler/intrinsics/string/TestStringLatin1IndexOfChar.java (hunk 4 above)
Comment (jasontatton-aws) [reply test/hotspot/jtreg/compiler/intrinsics/string/TestStringLatin1IndexOfChar.java:1]: Thanks I have added this now

File: test/micro/org/openjdk/bench/java/lang/StringIndexOfChar.java (hunk 5 above)
Comment (jasontatton-aws) [reply test/micro/org/openjdk/bench/java/lang/StringIndexOfChar.java:23]: Thanks I have added this now

**Review by neliasso — APPROVED** [review]

**Comment by RogerRiggs** [comment]
Due to the requirement for multiple reviewers, I had been waiting to add my review of the Core-Libs files until the HotSpot reviewers had approved!

**Comment by RogerRiggs** [comment]
This integration without testing with a current merge from the master and has caused two build failures.
JDK-8254761: Wrong intrinsic annotation used for StringLatin1.indexOfChar

Also, there is a raw unicode character in the JMH test that causes a compilation error.
```== Output from failing command(s) repeated here ===
[2020-10-14T14:39:09,608Z] * For target support_test_micro_classes__the.BUILD_JDK_MICROBENCHMARK_batch:
[2020-10-14T14:39:09,611Z] /opt/mach5/mesos/work_dir/slaves/4076d11c-c6ed-4d07-84c1-4ab8d55cd975-S108796/frameworks/1735e8a2-a1db-478c-8104-60c8b0af87dd-0196/executors/400e1f56-2d49-42d8-8879-97d4fbb6c909/runs/c49da2bc-a8fe-4a5d-8159-57a9b0316fd2/workspace/open/test/micro/org/openjdk/bench/java/lang/StringIndexOfChar.java:71: error: unmappable character (0xE2) for encoding ascii
[2020-10-14T14:39:09,611Z]             sb.append(isUtf16?'???':'b');
[2020-10-14T14:39:09,611Z]                                ^
[2020-10-14T14:39:09,611Z] /opt/mach5/mesos/work_dir/slaves/4076d11c-c6ed-4d07-84c1-4ab8d55cd975-S108796/frameworks/1735e8a2-a1db-478c-8104-60c8b0af87dd-0196/executors/400e1f56-2d49-42d8-8879-97d4fbb6c909/runs/c49da2bc-a8fe-4a5d-8159-57a9b0316fd2/workspace/open/test/micro/org/openjdk/bench/java/lang/StringIndexOfChar.java:71: error: unmappable character (0x98) for encoding ascii
[2020-10-14T14:39:09,611Z]             sb.append(isUtf16?'???':'b');
[2020-10-14T14:39:09,611Z]                                 ^
[2020-10-14T14:39:09,611Z] /opt/mach5/mesos/work_dir/slaves/4076d11c-c6ed-4d07-84c1-4ab8d55cd975-S108796/frameworks/1735e8a2-a1db-478c-8104-60c8b0af87dd-0196/executors/400e1f56-2d49-42d8-8879-97d4fbb6c909/runs/c49da2bc-a8fe-4a5d-8159-57a9b0316fd2/workspace/open/test/micro/org/openjdk/bench/java/lang/StringIndexOfChar.java:71: error: unmappable character (0xBA) for encoding ascii
[2020-10-14T14:39:09,611Z]             sb.append(isUtf16?'???':'b');
[2020-10-14T14:39:09,611Z]                                  ^
[2020-10-14T14:39:09,611Z] /opt/mach5/mesos/work_dir/slaves/4076d11c-c6ed-4d07-84c1-4ab8d55cd975-S108796/frameworks/1735e8a2-a1db-478c-8104-60c8b0af87dd-0196/executors/400e1f56-2d49-42d8-8879-97d4fbb6c909/runs/c49da2bc-a8fe-4a5d-8159-57a9b0316fd2/workspace/open/test/micro/org/openjdk/bench/java/lang/StringIndexOfChar.java:71: error: unclosed character literal
[2020-10-14T14:39:09,611Z]             sb.append(isUtf16?'???':'b');
[2020-10-14T14:39:09,611Z]                               ^
[2020-10-14T14:39:09,611Z] /opt/mach5/mesos/work_dir/slaves/4076d11c-c6ed-4d07-84c1-4ab8d55cd975-S108796/frameworks/1735e8a2-a1db-478c-8104-60c8b0af87dd-0196/executors/400e1f56-2d49-42d8-8879-97d4fbb6c909/runs/c49da2bc-a8fe-4a5d-8159-57a9b0316fd2/workspace/open/test/micro/org/openjdk/bench/java/lang/StringIndexOfChar.java:71: error: unclosed character literal
[2020-10-14T14:39:09,611Z]             sb.append(isUtf16?'???':'b');```

**Comment by RogerRiggs** [comment]
And also a failed Graal test because of the new intrinsic.

And JDK-8254785: compiler/graalunit/HotspotTest.java failed with "missing Graal intrinsics for: java/lang/StringLatin1.indexOfChar([BIII)I"

@phohensee don't be so quick to type `/sponsor`; there are three separate build and test failures.

**Comment by dcubed-ojdk** [comment]
@phohensee - @vnkozlov has determined that new Tier2 test failure is
also caused by this fix. See https://bugs.openjdk.java.net/browse/JDK-8254790.

**Pull request closed without merging**